src/dac.rs

```diff
@@ -0,0 +1,220 @@
+                DacBits::TwelveR => self.regs.dhr12r1.modify(|_, w| unsafe { w.bits(val) }),
+            },
+            DacId::Two => match self.bits {
+                DacBits::EightR => self.regs.dhr8r2.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveL => self.regs.dhr12l2.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveR => self.regs.dhr12r2.modify(|_, w| unsafe { w.bits(val) }),
+            },
+        }
+    }
+
+    /// Set the DAC voltage. `v` is in Volts.
+    pub fn set_voltage(&mut self, volts: f32) {
+        let val = match self.bits {
+            DacBits::EightR => ((volts / self.vref) * 255.) as u32,
+            DacBits::TwelveL => ((volts / self.vref) * 4_095.) as u32,
+            DacBits::TwelveR => ((volts / self.vref) * 4_095.) as u32,
+        };
+
+        self.set_value(val);
+    }
+
+    // Select and activate a trigger. See f303 Reference manual, section 16.5.4.
+    pub fn set_trigger(&mut self, trigger: Trigger) {
+        match self.id {
+            DacId::One => {
+                self.regs.cr.modify(|_, w| w.ten1().set_bit());
+                self.regs
+                    .cr
+                    .modify(|_, w| unsafe { w.tsel1().bits(trigger.bits()) });
+            }
+            DacId::Two => {
+                self.regs.cr.modify(|_, w| w.ten2().set_bit());
+                self.regs.cr.modify(|_, w| w.tsel2().bits(trigger.bits()));
+            }
+        }
+    }
+
+    /// Independent trigger with single LFSR generation
+    /// See f303 Reference Manual section 16.5.2
+    pub fn trigger_lfsr(&mut self, trigger: Trigger, data: u32) {
+        // todo: This may not be correct.
+        match self.id {
+            DacId::One => {
+                self.regs.cr.modify(|_, w| unsafe { w.wave1().bits(0b01) });
+                self.regs.cr.modify(|_, w| w.mamp1().bits(0b01));
+            }
+            DacId::Two => {
+                self.regs.cr.modify(|_, w| unsafe { w.wave2().bits(0b01) });
+                self.regs.cr.modify(|_, w| w.mamp2().bits(0b01));
+            }
+        }
+        self.set_trigger(trigger);
+        self.set_value(data);
+    }
+
+    /// Independent trigger with single triangle generation
+    /// See f303 Reference Manual section 16.5.2
+    pub fn trigger_triangle(&mut self, trigger: Trigger, data: u32) {
+        // todo: This may not be correct.
+        match self.id {
+            DacId::One => {
+                self.regs.cr.modify(|_, w| unsafe { w.wave1().bits(0b10) });
+                self.regs.cr.modify(|_, w| w.mamp1().bits(0b10));
+            }
+            DacId::Two => {
+                self.regs.cr.modify(|_, w| unsafe { w.wave2().bits(0b10) });
+                self.regs.cr.modify(|_, w| w.mamp2().bits(0b10));
+            }
+        }
+        self.set_trigger(trigger);
+        self.set_value(data);
+    }
+}
+
+impl fmt::Debug for Dac {
```

Comment:
We usually don't impl the `Debug` trait for peripherals in the HAL. What's the reason for doing it in this case?

Reply:
If not really needed, try to avoid `fmt` alltogether, because [Formatting is Unreasonably Expensive for Embedded Rust](https://jamesmunns.com/blog/fmt-unreasonably-expensive/)